client/src/GameState.ts

```diff
@@ -49,16 +49,44 @@ export class GameState {
             new Tetromino(TetrominoType.T),
         ];
 
+        // initial rotation
+        this.otherPieces.map((tetro, i) => {
+            tetro.setRotatedPosition(tetro.position, i + 1);
+            tetro.setRotation(i + 1);
+        });
+
         this.socket.on("initPlayer", (playerId) => {
             this.playerId = playerId;
-            console.log("playerId: ", playerId);
         });
 
         this.socket.on("playerMove", (playerId, state) => {
             const i = (3 - this.playerId + playerId) % 4; // Circular distance
+            this.otherPieces[i].position = state.position; // position before rotation
             this.otherPieces[i].setType(state.type);
             this.otherPieces[i].setRotatedPosition(state.position, i + 1);
             this.otherPieces[i].setRotation(i + 1 + state.rotation);
         });
+
+        this.socket.on("playerPlace", (playerId, state) => {
+            if (playerId == this.playerId) {
+                // the placing event is emitted by this very client, who already handled the local board during emitting
+                return;
+            }
+
+            // place the tetro on our board
+            const i = (3 - this.playerId + playerId) % 4; // Circular distance
+            const tetroToPlace = Tetromino.createFromState(state);
+            // local rotate
+            tetroToPlace.setRotation(i + 1 + tetroToPlace.rotation);
+            tetroToPlace.setRotatedPosition(tetroToPlace.position, i + 1);
```

Comment:
What is the motivation for nullable `otherTetrominoes`? Don't we expect them to always have some value, except for when they are not connected?

I was also imagining this array of players to be modified periodically by dead reckoning (overwritten by real data), but that is something later.

Reply:
TL:DR ignore that suggestion, I had false concerns (oops!)

I think the initial thought came from the worry that we might place a tetromino multiple times if we haven't received updates from the other clients (because their pieces would just sit there), but I forgot that a client only handles placing for its own tetromino, so that's not an actual concern.

That thinking eventually lead to something along the lines of "we spawn a new tetromino at the top of their section after their last one gets placed but we don't know the type so would it just shift next time we get an update?"... but again, no. We don't spawn a tetromino for them, they just send us an update once they do.